locale.c: Change to use enum instead of int

khwilliamson opened the pull request:
Knowing that the domain is just a few values can lead the compiler to generate better code, and to catch programmer errors.

This also creates #defines for the interior of a C 'for' loop header, which I think are more readable.

khwilliamson committed:
locale.c: Change to use enum instead of int

Knowing that the domain is just a few values can lead the compiler to
generate better code, and to catch programmer errors.

This also creates #defines for the interior of a C 'for' loop header,
which I think are more readable.

ilmari commented on locale.c:
```diff
@@ -441,6 +441,22 @@ S_wsetlocale(const int category, const wchar_t * wlocale)
+    i < LC_ALL_INDEX_;                                                      \
+    i = (locale_category_index) ((int) i + 1)
+
+#define all_category_indexes(i)                                             \
+    locale_category_index i = (locale_category_index) 0;                    \
+    i <= LC_ALL_INDEX_;                                                     \
+    i = (locale_category_index) ((int) i + 1)
+
```
This looks to be redundant with `all_individual_category_indexes`, and unused.

khwilliamson replied:
Maybe the name should be changed to something like `for_every_category`.  The difference is that the upper limit is `<=` which includes `LC_ALL`, whereas the other includes only the individual categories, and the upper limit is `<`.  Future commits in the pipeline use this one

khwilliamson replied:
I'll come up with a better name when the first use of this macro is added

ilmari commented on locale.c:
```diff
@@ -441,6 +441,22 @@ S_wsetlocale(const int category, const wchar_t * wlocale)
 #  endif
 #endif  /* WIN32_USE_FAKE_OLD_MINGW_LOCALES */
 
+/* 'for' loop headers to hide the necessary casts */
```
Since these are only usable inside the control block of a `for` loop, I think it makes more sense to include that in the macro definition, e.g. 

```
#define for_all_individual_category_indexes(i) \
    for (locale_category_index i = … 
```

khwilliamson replied:
Done

khwilliamson committed:
locale.c: Include the 'for' in the macro name

This changes the previous commit so that instead of the interior of the
'for' header being replaced by a macro, the entire 'for (...)' is
replaced.

This could be squashed with the previous commit if desired.
